### ISSUE TITLE
truffleAssert problem

### ISSUE THREAD

**Issue by Aleione** [trufflesuite/truffle#2390]
Hi,

i have a problem with truffleAssert during truffle test. On local ganache blockchain test is passed because revert string is the same of that indicate in javascript test code. When i run test with the same code on ropsten using truffle-hdwallet-provider module, it returns this error:
 
AssertionError: Expected to fail with SafeMath: subtraction overflow, but failed with: StatusError: Transaction: 0x9b3b35d6d42be4c8fb17e8e7c540a4090a1d36e0ac7b60f5888662b3c9d19385 exited with an error (status 0).
     Please check that the transaction:
     - satisfies all conditions set by Solidity `require` statements.
     - does not trigger a Solidity `revert` statement.

It seems that revert error string is not that indicated in solidity code but a transaction error generic.
How can i fix this, please?

Thanks

**Comment by gnidan** [comment]
Hi @Aleione, thanks for raising this! We'll triage this and get back to you as soon as we can!

**Comment by haltman-at** [comment]
This sounds like it might be a problem with truffle-assertions rather than Truffle; @rkalis, do you know anything about this?  I do have to wonder whether truffle-assertions may be relying on special Ganache behavior here (I don't know anything about truffle-assertions' internals); providing revert strings for transactions (as opposed to calls) is nonstandard Ganache behavior that won't occur with other clients.  Truffle determines revert strings in other cases by rerunning the transaction as a call, but I don't know whether truffle-assertions is purely built on top of Truffle or whether it might go a little more low-level and get the revert strings directly from the Ganache error.  In the latter case, this will have problems with other clients.  Anyway, like I said, I'm hoping @rkalis can provide som clarity here.

**Comment by rkalis** [comment]
Hey @haltman-at, thanks for the ping.

[This](https://github.com/rkalis/truffle-assertions/blob/master/index.js#L146-L162) is the function in `truffle-assertions` that asserts a revert. [This](https://github.com/rkalis/truffle-assertions#truffleassertfailsasyncfn-errortype-reason-message) is the documentation for the function.

The function takes a Promise produced by a contract function on a TruffleContract instance. It then asserts that the Promise is rejected, and checks the rejection message (which includes the reason string when used with Ganache / ganache-cli). But when using public networks like Ropsten, the error is like in the issue description, and doesn't include the reason string:
```
StatusError: Transaction: 0x9b3b35d6d42be4c8fb17e8e7c540a4090a1d36e0ac7b60f5888662b3c9d19385 exited with an error (status 0).
```

You said that Truffle determines the revert reason string in those cases by rerunning the tx as a call. So to make use of that I should extend the `.fails()` function so that it reruns the tx as a call when it receives a `StatusError` to determine the reason string? Or is there something baked into Truffle that I can use for that?

**Comment by haltman-at** [comment]
So, I'm not the expert here; @eggplantzzz would know more I think.  But I think if you're doing things by means of Truffle, you ought to be able to make use of its doing that rather than having to do it yourself.  Except this might be quite troublesome because I think this gets thrown in an exception and it may not be a proper `Error` object and extracting it may require hackery?  Like I said, this is about the limit of my knowledge (at least offhand), hoping @eggplantzzz or someone can elaborate.

**Comment by eggplantzzz** [comment]
Hey @cgewecke, do you happen to know offhand if Truffle provides a way for truffle-assertions to get at the revert string for this?

**Comment by cgewecke** [comment]
@eggplantzzz (Just speculating here..)

The truffleAssert code LGTM. I suspect this might be a bug in the way TC handles reason fetching from geth-like clients when gas estimation is enabled. There's a related issue at [ganache-core 480][2] - they're behavior became more geth-like last month.

@Aleione is using Ropsten via HDWalletProvider.
+ Ropsten is geth and there are reason fetching tests for geth at truffle/contract which pass [here][1]
+ They don't work if gas estimation is enabled because Geth throws the error on revert instead of returning an estimate and as written TC [passes 0 gas][3] for that case. (That design isn't great -mea culpa). 

A quick test for this hypothesis would be: does the reason get fetched from Ropsten when you fund the tx with an explicit gas value (like TC's unit test)?

I don't know why this isn't OOG tbh.

[1]: https://github.com/trufflesuite/truffle/blob/develop/packages/contract/test/methods.js#L465-L486
[2]: https://github.com/trufflesuite/ganache-core/issues/480
[3]: https://github.com/trufflesuite/truffle/blob/develop/packages/contract/lib/execute.js#L19-L46

**Comment by gnidan** [comment]
Hm so what do we do here? @haltman-at says that this is a truffle-assertions issue, not a Truffle issue, which I'm inclined to believe... although there seems to be more that Truffle could do to make truffle-assertions' life easier?

Thoughts for resolving this, anyone?

**Comment by cgewecke** [comment]
> does the reason get fetched from Ropsten when you fund the tx with an explicit gas value (like [TC's unit test](https://github.com/trufflesuite/truffle/blob/develop/packages/contract/test/methods.js#L470))

My belief is that's a reasonable test of whether or not it's a Truffle gas estimation error handling problem. If the reason **is fetched** - it's Truffle's problem.

**Comment by rkalis** [comment]
Hmm. I did some digging into this, and tested some of the things mentioned here. Doing so, I noticed though that this actually just seems to work out of the box.

Considering the following ([source](https://github.com/rkalis/truffle-assertions/blob/master/docs/kalis-me-tutorial-code/test/test-reverts.test.js#L26-L29)): 
```js
await truffleAssert.reverts(
  casino.bet(betNumber, { from: bettingAccount, value: betSize }),
  "Bet amount can not exceed max bet size"
);
```

When following these steps in my [tutorial code project](https://github.com/rkalis/truffle-assertions/tree/master/docs/kalis-me-tutorial-code):
1. `truffle compile`
2. `truffle migrate --network rinkeby` or `truffle migrate --network goerli` or `truffle migrate --network ropsten`
3. `truffle test --network rinkeby` or `truffle test --network goerli` or `truffle test --network ropsten`

All tests pass
```
  Contract: Casino
    ✓ should not be able to bet more than max bet (15417ms)
    ✓ should not be able to bet without sending ether (14403ms)
```

This is without specifying any additional gas parameter, just out of the box. It also works fine when using a gas parameter, but it doesn't seem to matter.

@Aleione do you have a repository with the code you are using so we can try to reproduce the behaviour you're experiencing, as I can't reproduce it with my own tests.

## Environment
* Operating System: macOS 10.14.4
* Truffle version: 5.0.38
* node version: 11.15.0
* npm version: 6.9.0

**Comment by gnidan** [comment]
@rkalis our ticket processing process says it's time to close this from no response, but want to leave it up to you... what do we do here, do you think?

**Comment by rkalis** [comment]
Haven't heard back from @Aleione, and I couldn't reproduce the issue without their response, so closing is fine by me. It can always be reopened with new information in the future.

**Comment by gnidan** [comment]
Sounds good, thanks @rkalis! Doing that!
